### ISSUE TITLE
False positives `Avoid defining a one-member abstract class when a simple function will do

### ISSUE THREAD

**Issue by zoechi** [dart-lang/linter#64]
/lib/src/connection/connection.dart 50:16 [lint] Avoid defining a one-member abstract class when a simple function will do.
abstract class Config {
               ^^^^^^
```dart
abstract class Config {
  String get datasetId;
}
```
I think when the class has at least one unimplemented getter/setter/method this rule doesn't apply anymore
I'm not sure what this rule is about exactly but I guess this also only applies when it has only static members.

**Comment by pq** [comment]
Hmmm.  The rationale for this lint is to discourage classes like 

```
class Predicate {
  bool test();
}
```

where a `typedef` would arguably be better.

That said, your example doesn't fit this pattern at all.  I'll look at tightening up the logic.

Thanks!

**Comment by pq** [comment]
Edit: OK so strictly speaking, it does fit the bill.  For further reference, here's where it's discussed in the [style guide](https://www.dartlang.org/articles/style-guide/#types).  I'll mull over how we might improve it's logic.  It may just be hard to not trigger false positives since it's hard to infer the intent statically.

**Comment by pq** [comment]
cc @scheglov and @bwilkerson who may have thoughts on improving the logic.

**Comment by bwilkerson** [comment]
> ... it's hard to infer the intent statically ...

That's an understatement! It's nearly (or maybe completely) impossible to infer intent.

But sometimes you can use convention to make it appear that you understood the intent. In this case, I'd recommend limiting the rule to abstract classes that define a single abstract method. If they define a single abstract getter or setter, then don't produce a lint.

The rationale behind this is as follows. Getters and setters are designed to be used to give the appearance of the class having a property. Saying that an instance of the class has a property is roughly the same as saying that it has state. Given that functions don't have state, a class with state cannot have been intended to be equivalent to a function. Therefore, it would not work to replace such a class with a function type alias.

**Comment by pq** [comment]
> That's an understatement!

Ha!  Well, yeah. :)

As for your refinement Brian, I buy it.  I think that will rule out a bunch of false positives (this one included).  In the end, due to the nature of the advice, I think this is one of those rules that may just not always apply.  This fix will help a lot though.  Thanks!

**Comment by Pajn** [comment]
I don't understand this lint error.

I may want interfaces with just one method even if the class itself contain more than that method.

In my case I have this abstract class
```dart
abstract class ToAction {
  bool deliverTo(user);
}
```
It's used in a Flux pattern where actions are instances of classes.
Actions which should be sent to a user implements that class and implements the `deliverTo` method to filter which users the action should be delivered to.

For example:
```dart
class NewMessage implements ToAction {
  final Message message;

  NewMessage(this.message);

  @override
  bool deliverTo(user) => message.to == user;
}
```

If I would do this with typedefs instead I would need both more and uglier code:
```dart
typedef bool DeliverTo(user);

abstract class ToAction {
  DeliverTo get deliverTo;
}

class NewMessage extends ToAction {
  final Message message;

  @override
  DeliverTo get deliverTo => _deliverTo;

  NewMessage(this.message);

  bool _deliverTo(user) => message.to == user;
}
```

And if I want `ToAction` generic I don't think that's possible with typedefs.
```dart
abstract class ToAction<T> {
  bool deliverTo(T user);
}
```

The style guide says one should avoid that pattern when a typedef would do, but in the described use case it wouldn't because I need the object (because I need the public message field). However if my concrete class had no other public fields than my function, then a typedef would do as a plain function would do.

Because the linter can't see on the abstract class which of the uses I intend I think the error should be changed to warn on concrete classes with only a single method (and no public properties). As of now it will report false positives.
